photon-lib/src/main/java/org/photonvision/simulation/VideoSimUtil.java

```diff
@@ -0,0 +1,335 @@
+            Imgproc.resize(
+                    tagImage, scaledTagImage, new Size(), scaleFactor, scaleFactor, Imgproc.INTER_NEAREST);
+            tagPoints.fromArray(get16h5MarkerPts(scaleFactor));
+        } else tagImage.assignTo(scaledTagImage);
+
+        // constrain the bounding rect inside of the destination image
+        boundingRect.x -= 1;
+        boundingRect.y -= 1;
+        boundingRect.width += 2;
+        boundingRect.height += 2;
+        if (boundingRect.x < 0) {
+            boundingRect.width += boundingRect.x;
+            boundingRect.x = 0;
+        }
+        if (boundingRect.y < 0) {
+            boundingRect.height += boundingRect.y;
+            boundingRect.y = 0;
+        }
+        boundingRect.width = Math.min(destination.width() - boundingRect.x, boundingRect.width);
+        boundingRect.height = Math.min(destination.height() - boundingRect.y, boundingRect.height);
+        if (boundingRect.width <= 0 || boundingRect.height <= 0) return;
+
+        // upscale if supersampling
+        Mat scaledDstPts = new Mat();
+        if (supersampling > 1) {
+            Core.multiply(dstPoints, new Scalar(supersampling, supersampling), scaledDstPts);
+            boundingRect.x *= supersampling;
+            boundingRect.y *= supersampling;
+            boundingRect.width *= supersampling;
+            boundingRect.height *= supersampling;
+        } else dstPoints.assignTo(scaledDstPts);
+
+        // update transform relative to expanded, scaled bounding rect
+        Core.subtract(scaledDstPts, new Scalar(boundingRect.tl().x, boundingRect.tl().y), scaledDstPts);
+        perspecTrf = Imgproc.getPerspectiveTransform(tagPoints, scaledDstPts);
+
+        // warp (scaled) tag image onto (scaled) ROI image representing the portion of
+        // the destination image encapsulated by boundingRect
+        Mat tempROI = new Mat();
+        Imgproc.warpPerspective(scaledTagImage, tempROI, perspecTrf, boundingRect.size(), warpStrategy);
+
+        // downscale ROI with interpolation if supersampling
+        if (supersampling > 1) {
+            boundingRect.x /= supersampling;
+            boundingRect.y /= supersampling;
+            boundingRect.width /= supersampling;
+            boundingRect.height /= supersampling;
+            Imgproc.resize(tempROI, tempROI, boundingRect.size(), 0, 0, Imgproc.INTER_AREA);
+        }
+
+        // we want to copy ONLY the transformed tag to the result image, not the entire bounding rect
+        // using a mask only copies the source pixels which have an associated non-zero value in the
+        // mask
+        Mat tempMask = Mat.zeros(tempROI.size(), CvType.CV_8UC1);
+        Core.subtract(
+                extremeCorners, new Scalar(boundingRect.tl().x, boundingRect.tl().y), extremeCorners);
+        Point tempCenter = new Point();
+        tempCenter.x =
+                Arrays.stream(extremeCorners.toArray()).mapToDouble(p -> p.x).average().getAsDouble();
```

Comment:
is this confirmed to be not empty? average returns an optional double

Reply:
`extremeCorners` should only be empty if the tag image is empty/null, which is handled.